src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1133,64 +1163,83 @@
                   previously-reduced? (not= :clara.rules.memory/no-accum-reduced previous-candidates)
                   previous-candidates (when previously-reduced? previous-candidates)]]
 
-      ;; Combine the newly reduced values with any previous items.
-      (let [combined-candidates (into previous-candidates candidates)]
+      ;; Combine the newly reduced values with any previous items.  Ensure that new items are always added to the end so that
+      ;; we have a consistent order for retracting results from accumulators such as acc/all whose results can be in any order.  Making this
+      ;; ordering consistent allows us to skip the filter step on previous elements on right-activations.
+      (let [combined-candidates (into []
+                                      cat
+                                      [previous-candidates candidates])]
 
         (l/add-accum-reduced! listener node join-bindings combined-candidates bindings)
 
-        (mem/add-accum-reduced! memory node join-bindings combined-candidates bindings)
-        (doseq [token matched-tokens
+        (mem/add-accum-reduced! memory node join-bindings combined-candidates bindings))
+      
+      (doseq [token matched-tokens
+
+              :let [previous-filtered-facts (filter-accum-facts join-filter-fn token previous-candidates)
+
+                    new-filtered-facts (filter-accum-facts join-filter-fn token candidates)]
+
+              ;; If no new elements matched the token, we don't need to do anything for this token
+              ;; since the final result is guaranteed to be the same.
+              :when (seq new-filtered-facts)
+
+              :let [previous-accum-result-init (cond
+                                                 (seq previous-filtered-facts)
+                                                 (do-accumulate accumulator previous-filtered-facts)
+
+                                                 (and (-> accumulator :initial-value some?)
+                                                      (empty? new-bindings))
+                                                 (:initial-value accumulator)
+
+                                                 ;; Allow direct determination later of whether there was a previous value
+                                                 ;; as determined by the preceding cond conditions.
+                                                 :else ::no-previous-value)
```

Comment:
Why can't this case just be
```clj
:else
(:initial-value accumulator)
```
This special case seems hard to follow for me and it looks like it is being checked several times later too.